src/Microsoft.AspNetCore.Razor.Language/VirtualRazorProjectFileSystem.cs

```diff
@@ -0,0 +1,200 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal class VirtualRazorProjectFileSystem : RazorProjectFileSystem
+    {
+        private readonly DirectoryNode Root = new DirectoryNode("/", "/");
+
+        public override IEnumerable<RazorProjectItem> EnumerateItems(string basePath)
+        {
+            basePath = NormalizeAndEnsureValidPath(basePath);
+            var directory = Root.GetDirectory(basePath);
+            return directory?.EnumerateItems() ?? Enumerable.Empty<RazorProjectItem>();
+        }
+
+        public override RazorProjectItem GetItem(string path)
+        {
+            path = NormalizeAndEnsureValidPath(path);
+            return Root.GetItem(path) ?? new NotFoundProjectItem(string.Empty, path);
+        }
+
+        public void Add(RazorProjectItem projectItem)
+        {
+            var filePath = NormalizeAndEnsureValidPath(projectItem.FilePath);
+            Root.AddItem(new FileNode(filePath, projectItem));
+        }
+
+        private class DirectoryNode
```

Comment:
You should break this out into its own type (or expose internal if you don't want to do that) and test it separately. It has a lot of logic which I'm uncomfortable relying on integration level tests to verify.

Reply:
Sure